### PR TITLE
bazel: Shift dependency validate tool to bazel

### PR DESCRIPTION
**Pull request by phlax**
Commit Message:

this is a reopening of #18085 partly for the purposes of testing

im not sure whether it would be best to shift this tool to bazel first and then to `envoy.dependency.check` or just to the check tool directly

either way i need to figure out separating the data

Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: tools/dependency/validate.py [hunk 1]
```diff
@@ -122,28 +107,51 @@ def query_external_deps(self, *targets):
...
-                # it transitively.
-                if ext_dep in self._implied_untracked_deps_revmap:
-                    ext_dep = self._implied_untracked_deps_revmap[ext_dep]
-                ext_deps.add(ext_dep)
+            if not d:
+                continue
+            try:
+                ext_dep = BAZEL_QUERY_EXTERNAL_DEP_RE.match(d).group(1)
+            except:
+                print(f"BAD MATCH: {d}")
```
Comment (phlax) [inline tools/dependency/validate.py:146]: just continue on no match

Comment (htuch) [reply tools/dependency/validate.py:146]: Seems a pretty broad exception catch, can this be made more specific?

Comment (phlax) [reply tools/dependency/validate.py:146]: checking the original code it was "look before you leap", so ill just get it to continue if `match` is `None`

File: tools/dependency/BUILD [hunk 2]
```diff
@@ -41,3 +41,25 @@ py_binary(
...
+    data = [
+        "//bazel:all_repository_locations",
+        "//source/extensions:extensions_build_config",
+    ],
+    deps = [
+        "//tools/base:bazel_query",
+    ],
+)
+
+py_binary(
```
Comment (htuch) [inline tools/dependency/BUILD:61]: There isn't `py_test`?

File: tools/dependency/validate.py [hunk 3]
```diff
@@ -122,28 +107,51 @@ def query_external_deps(self, *targets):
...
-            if match:
-                ext_dep = match.group(1)
-                if ext_dep in self._ignore_deps:
-                    continue
-                # If the dependency is untracked, add the source dependency that loaded
-                # it transitively.
-                if ext_dep in self._implied_untracked_deps_revmap:
-                    ext_dep = self._implied_untracked_deps_revmap[ext_dep]
-                ext_deps.add(ext_dep)
+            if not d:
```
Comment (htuch) [inline tools/dependency/validate.py:141]: How is this possible?

File: tools/dependency/validate.py (hunk 1 above)
Comment (htuch) [reply tools/dependency/validate.py:146]: Is this just an assert condition?

**Review by htuch — COMMENTED** [review]
LGTM modulo nits.

File: tools/dependency/validate.py (hunk 1 above)
Comment (phlax) [reply tools/dependency/validate.py]: ive made this just do it what it did before

File: tools/dependency/validate.py (hunk 3 above)
Comment (phlax) [reply tools/dependency/validate.py:141]: i think this was mistakenly trying to filter out the `match` that was not being filtered out previously, removed

**Comment by phlax** [comment]
This one is difficult

The recursive bazel queries that it does `//source/...` etc pull in huge amounts of transitive dependencies and cant be shifted to a `genquery`

I think we want to separate out the data from the processing further (ultimately i want to shift this to envoy.dependency.check so we kinda want to boil the data down to json somehow) - but i think this is a good first step at least just getting the tool into bazel

I wanted to be confident that it was checking the same thing so i added some `len()` s around query results in the logging, and compared this to current #19568 

im pretty confident it does the right thing

**Comment by phlax** [comment]
re optimization - i think its probably not faster (maybe slightly slower) but it uses a lot less memory by using exclude clauses in the bazel queries, rather than getting both inclusive and exclusive sets, and then filtering in python

File: tools/dependency/BUILD (hunk 2 above)
Comment (phlax) [reply tools/dependency/BUILD:53]: this has the existing `validate_test` script set up - below and in ci (ci/do_ci.sh) so its testing the same as now i think

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**